### PR TITLE
resolved: three tweaks to server feature level tracking

### PR DESCRIPTION
**Pull request by poettering**

### CODE REVIEW

**Comment by poettering** [comment]
Split out of #17535

File: src/resolve/resolved-dns-server.c [hunk 1]
```diff
@@ -488,42 +495,57 @@ DnsServerFeatureLevel dns_server_possible_feature_level(DnsServer *s) {
...
 
                 } else if (s->n_failed_udp >= DNS_SERVER_FEATURE_RETRY_ATTEMPTS &&
-                           s->possible_feature_level >= (dns_server_get_dnssec_mode(s) == DNSSEC_YES ? DNS_SERVER_FEATURE_LEVEL_LARGE : DNS_SERVER_FEATURE_LEVEL_UDP)) {
+                           DNS_SERVER_FEATURE_LEVEL_IS_UDP(s->possible_feature_level) &&
+                           ((s->possible_feature_level != DNS_SERVER_FEATURE_LEVEL_DO) || dns_server_get_dnssec_mode(s) != DNSSEC_YES)) {
 
-                        /* We lost too many UDP packets in a row, and are on a feature level of UDP or higher. If the
-                         * packets are lost, maybe the server cannot parse them, hence downgrading sounds like a good
-                         * idea. We might downgrade all the way down to TCP this way.
+                        /* We lost too many UDP packets in a row, and are on a UDP feature level of UDP. If
```
Comment (yuwata) [inline src/resolve/resolved-dns-server.c]: 'are on a UDP feature level of UDP' sounds redundant for me.

**Review by yuwata — APPROVED** [review]
LGTM. One minor comment.

**Comment by poettering** [comment]
fixed the issue @yuwata found and rebased. no other changes. upgrading green label.

**Pull request merged**